stellargraph/core/element_data.py

```diff
@@ -335,29 +381,62 @@ def __init__(self, ids, sources, targets, weights, type_starts):
         self._empty_ilocs = np.array([], dtype=np.uint8)
 
     def _init_directed_adj_lists(self):
-        # record the edge ilocs of incoming, outgoing and both-direction edges
-        in_dict = {}
-        out_dict = {}
-
-        for i, (src, tgt) in enumerate(zip(self.sources, self.targets)):
-            in_dict.setdefault(tgt, []).append(i)
-            out_dict.setdefault(src, []).append(i)
-
-        dtype = np.min_scalar_type(len(self.sources))
-        self._edges_in_dict = _numpyise(in_dict, dtype=dtype)
-        self._edges_out_dict = _numpyise(out_dict, dtype=dtype)
+        # record the edge ilocs of incoming and outgoing edges
+        if len(self.targets) == 0:
+            empty_array = np.array([], dtype=np.uint8)
+            self._edges_in_dict = FlatAdjacencyList(empty_array, empty_array)
+            self._edges_out_dict = FlatAdjacencyList(empty_array, empty_array)
+            return
+
+        def _to_dir_adj_list(arr):
+            neigh_counts = np.bincount(arr, minlength=self.number_of_nodes)
+            neigh_counts = neigh_counts.astype(self._id_index.dtype, copy=False)
+            splits = np.cumsum(neigh_counts, dtype=self._id_index.dtype)
+            flat = np.argsort(arr).astype(self._id_index.dtype, copy=False)
+            return FlatAdjacencyList(flat, splits)
+
+        self._edges_in_dict = _to_dir_adj_list(self.targets)
+        self._edges_out_dict = _to_dir_adj_list(self.sources)
 
     def _init_undirected_adj_lists(self):
```

Comment:
The fact we have to implement things both for directed mode and undirected mode led me to file https://github.com/stellargraph/stellargraph/issues/1426. Given this experience, you have any thoughts on that issue?